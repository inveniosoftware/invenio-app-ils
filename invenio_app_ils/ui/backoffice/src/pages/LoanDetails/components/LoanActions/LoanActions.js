import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { List, Button } from 'semantic-ui-react';

export default class LoanActions extends Component {
  constructor(props) {
    super(props);
    this.performLoanAction = this.props.performLoanAction;
  }

  renderAvailableActions(pid, loan, actions = {}) {
    return Object.keys(actions).map(action => {
      return (
        <List.Item key={action}>
          <Button
            primary
            onClick={() => {
              this.performLoanAction(pid, loan, actions[action]);
            }}
          >
            {action}
          </Button>
        </List.Item>
      );
    });
  }

  render() {
    const {
      availableActions: actions,
      metadata: loan,
      id: pid,
    } = this.props.loanDetails;
    return (
      <List horizontal>{this.renderAvailableActions(pid, loan, actions)}</List>
    );
  }
}

LoanActions.propTypes = {
  loanDetails: PropTypes.object.isRequired,
  performLoanAction: PropTypes.func,
};
