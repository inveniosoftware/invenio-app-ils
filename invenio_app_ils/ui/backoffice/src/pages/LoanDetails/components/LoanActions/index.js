export { default as LoanActions } from './LoanActions';
