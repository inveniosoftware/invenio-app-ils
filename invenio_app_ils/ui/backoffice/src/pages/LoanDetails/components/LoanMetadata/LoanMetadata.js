import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Grid, Segment, Form } from 'semantic-ui-react';

export class LoanMetadata extends Component {
  renderLoanMetadata(data) {
    return Object.keys(data.metadata).map(key => {
      return (
        <Form.Field key={key} name="loan-field">
          <label className="field-name">{key}</label>
          <p>{data.metadata[key]}</p>
        </Form.Field>
      );
    });
  }

  render() {
    let { data } = this.props;
    return (
      <Segment>
        <Grid>
          <Grid.Column>
            <h1>Loan - {data.metadata.loan_pid}</h1>
            <Form>{this.renderLoanMetadata(data)}</Form>
          </Grid.Column>
        </Grid>
      </Segment>
    );
  }
}

LoanMetadata.propTypes = {
  data: PropTypes.object.isRequired,
};
