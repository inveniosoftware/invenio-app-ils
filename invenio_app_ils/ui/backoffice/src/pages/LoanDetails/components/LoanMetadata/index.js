export { default as LoanMetadata } from './LoanMetadata';
