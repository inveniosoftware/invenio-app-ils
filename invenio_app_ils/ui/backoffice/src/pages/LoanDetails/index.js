import { compose } from 'redux';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { fetchLoanDetails, postLoanAction } from './state/actions';
import LoanDetailsContainerComponent from './LoanDetailsContainer';

const mapStateToProps = state => ({
  data: state.loanDetails.data,
  error: state.loanDetails.error,
  isLoading: state.loanDetails.isLoading,
  actionLoading: state.loanDetails.actionLoading,
  loanActionError: state.loanDetails.loanActionError,
});

const mapDispatchToProps = dispatch => ({
  fetchLoanDetails: loanId => dispatch(fetchLoanDetails(loanId)),
  postLoanAction: (url, data) => dispatch(postLoanAction(url, data)),
});

export const LoanDetailsContainer = compose(
  withRouter,
  connect(
    mapStateToProps,
    mapDispatchToProps
  )
)(LoanDetailsContainerComponent);
