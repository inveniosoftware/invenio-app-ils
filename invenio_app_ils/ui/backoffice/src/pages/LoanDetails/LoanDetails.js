import React, { Component } from 'react';
import { Loader, Message } from 'semantic-ui-react';
import { LoanTitle } from './components/LoanTitle/LoanTitle';
import { LoanMetadata } from './components/LoanMetadata/LoanMetadata';
import { LoanActions } from './components/LoanActions/LoanActions';

export default class LoanDetails extends Component {
  constructor(props) {
    super(props);
    this.fetchLoanDetails = this.props.fetchLoanDetails;
    this.postLoanAction = this.props.postLoanAction;
  }

  componentDidMount() {
    this.unlisten = this.props.history.listen((location, action) => {
      if (location.state && location.state.recid) {
        this.fetchLoanDetails(location.state.loanId);
      }
    });
    this.fetchLoanDetails(this.props.match.params.loanId);
  }

  componentWillUnmount() {
    this.unlisten();
  }

  render() {
    let {
      data,
      error,
      fetchLoading,
      actionLoading,
      loanActionError,
    } = this.props;
    if (fetchLoading) return <Loader active inline="centered" />;
    if (loanActionError) {
      return (
        <Message
          icon="exclamation"
          header="Oups, something went horribly wrong!"
          content={error.message}
        />
      );
    }
    console.log(data);
    return (
      <section>
        <LoanTitle loanId={data.metadata.loan_pid} />
        <LoanMetadata data={data} />
        <LoanActions
          data={data.metadata}
          actions={data.availableActions}
          actionLoading={actionLoading}
          onAction={this.postLoanAction}
        />
      </section>
    );
  }
}
