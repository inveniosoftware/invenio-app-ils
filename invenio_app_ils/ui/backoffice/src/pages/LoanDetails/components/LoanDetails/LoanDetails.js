import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Container } from 'semantic-ui-react';
import { compose } from 'redux';
import { withError, withLoader } from 'common/hoc';
import { LoanMetadata } from '../LoanMetadata/LoanMetadata';
import { LoanActions } from '../LoanActions/LoanActions';

class LoanDetails extends Component {
  render() {
    let { data, onAction } = this.props;
    return (
      <Container>
        <LoanMetadata data={data} />
        <LoanActions data={data} onAction={onAction} />
      </Container>
    );
  }
}

LoanDetails.propTypes = {
  data: PropTypes.object.isRequired,
};

export default compose(
  withLoader,
  withError
)(LoanDetails);
