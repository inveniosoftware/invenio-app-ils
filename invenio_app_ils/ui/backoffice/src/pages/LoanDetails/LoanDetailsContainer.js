import React, { Component } from 'react';
import PropTypes from 'prop-types';
import LoanDetails from './components/LoanDetails/LoanDetails';

export default class LoanDetailsContainer extends Component {
  constructor(props) {
    super(props);
    this.fetchLoanDetails = this.props.fetchLoanDetails;
    this.postLoanAction = this.props.postLoanAction;
  }

  componentDidMount() {
    this.unlisten = this.props.history.listen((location, action) => {
      if (location.state && location.state.loanid) {
        this.fetchLoanDetails(location.state.loanId);
      }
    });
    this.fetchLoanDetails(this.props.match.params.loanId);
  }

  componentWillUnmount() {
    this.unlisten();
  }

  render() {
    let { data, isLoading, error } = this.props;
    return (
      <LoanDetails
        data={data}
        isLoading={isLoading}
        onAction={this.postLoanAction}
        error={error}
      />
    );
  }
}

LoanDetailsContainer.propTypes = {
  isLoading: PropTypes.bool.isRequired,
};
