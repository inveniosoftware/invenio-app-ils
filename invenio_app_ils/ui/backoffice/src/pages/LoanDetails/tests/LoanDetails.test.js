import React from 'react';
import { mount } from 'enzyme';
import { Message, Loader } from 'semantic-ui-react';
import LoanDetails from '../LoanDetails';
import { LoanMetadata } from '../components/LoanMetadata';
import { LoanActions } from '../components/LoanActions';

const LOAN_DETAILS_DEFAULT_PROPS = {
  data: {},
  error: {},
  fetchLoading: true,
  actionLoading: false,
  loanActionError: false,
  fetchLoanDetails: () => 'fetchLoanDetails',
  postLoanAction: () => 'postLoanAction',
};

describe('tests the LoanDetails component', () => {
  let component;

  beforeAll(() => {
    LoanDetails.prototype.componentDidMount = () => {
      // omit `componentDidMount` to skip routing testing
    };
    LoanDetails.prototype.componentWillUnmount = () => {
      // omit `componentDidMount` to skip routing testing
    };
  });

  afterEach(() => {
    component.unmount();
  });

  it('renders initial state and checks the props are passed', () => {
    component = mount(<LoanDetails {...LOAN_DETAILS_DEFAULT_PROPS} />);
    expect(component).toMatchSnapshot();

    Object.keys(LOAN_DETAILS_DEFAULT_PROPS).forEach(key =>
      expect(component.props()[key]).toEqual(LOAN_DETAILS_DEFAULT_PROPS[key])
    );
  });

  it('renders the loader component when fetching is in progress', () => {
    component = mount(<LoanDetails {...LOAN_DETAILS_DEFAULT_PROPS} />);
    expect(component).toMatchSnapshot();

    const loader = component.find(Loader);
    const metadata = component.find(LoanMetadata);
    const actions = component.find(LoanActions);
    // should render Loader component only
    expect(loader.exists()).toEqual(true);
    expect(metadata.exists()).toEqual(false);
    expect(actions.exists()).toEqual(false);
  });

  it('renders a message when fetching fails', () => {
    let props = {
      ...LOAN_DETAILS_DEFAULT_PROPS,
      fetchLoading: false,
      loanActionError: true,
    };
    component = mount(<LoanDetails {...props} />);
    expect(component).toMatchSnapshot();

    const loader = component.find(Loader);
    const metadata = component.find(LoanMetadata);
    const actions = component.find(LoanActions);
    const message = component.find(Message);
    // should render Message component only
    expect(loader.exists()).toEqual(false);
    expect(metadata.exists()).toEqual(false);
    expect(actions.exists()).toEqual(false);
    expect(message.exists()).toEqual(true);
  });

  it('renders the details page when fetching succeeds', () => {
    let props = {
      ...LOAN_DETAILS_DEFAULT_PROPS,
      data: {
        metadata: {
          title: 'title',
        },
        availableActions: {
          action: 'action',
        },
      },
      error: {
        message: 'error',
      },
      fetchLoading: false,
    };
    component = mount(<LoanDetails {...props} />);
    expect(component).toMatchSnapshot();

    const loader = component.find(Loader);
    const metadata = component.find(LoanMetadata);
    const actions = component.find(LoanActions);
    const message = component.find(Message);
    // should render Message component only
    expect(loader.exists()).toEqual(false);
    expect(message.exists()).toEqual(false);
    expect(metadata.exists()).toEqual(true);
    expect(actions.exists()).toEqual(true);
  });
});
