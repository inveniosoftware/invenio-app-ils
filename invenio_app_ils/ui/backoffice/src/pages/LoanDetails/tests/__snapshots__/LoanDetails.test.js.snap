// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests the LoanDetails component renders a message when fetching fails 1`] = `
<LoanDetails
  actionLoading={false}
  data={Object {}}
  error={Object {}}
  fetchLoading={false}
  fetchLoanDetails={[Function]}
  loanActionError={true}
  postLoanAction={[Function]}
>
  <Message
    header="Oups, something went horribly wrong!"
    icon="exclamation"
  >
    <div
      className="ui icon message"
    >
      <Icon
        as="i"
        name="exclamation"
      >
        <i
          aria-hidden="true"
          className="exclamation icon"
        />
      </Icon>
      <MessageContent>
        <div
          className="content"
        >
          <MessageHeader
            content="Oups, something went horribly wrong!"
          >
            <div
              className="header"
            >
              Oups, something went horribly wrong!
            </div>
          </MessageHeader>
        </div>
      </MessageContent>
    </div>
  </Message>
</LoanDetails>
`;

exports[`tests the LoanDetails component renders initial state and checks the props are passed 1`] = `
<LoanDetails
  actionLoading={false}
  data={Object {}}
  error={Object {}}
  fetchLoading={true}
  fetchLoanDetails={[Function]}
  loanActionError={false}
  postLoanAction={[Function]}
>
  <Loader
    active={true}
    inline="centered"
  >
    <div
      className="ui active centered inline loader"
    />
  </Loader>
</LoanDetails>
`;

exports[`tests the LoanDetails component renders the details page when fetching succeeds 1`] = `
<LoanDetails
  actionLoading={false}
  data={
    Object {
      "availableActions": Object {
        "action": "action",
      },
      "metadata": Object {
        "title": "title",
      },
    }
  }
  error={
    Object {
      "message": "error",
    }
  }
  fetchLoading={false}
  fetchLoanDetails={[Function]}
  loanActionError={false}
  postLoanAction={[Function]}
>
  <Grid
    centered={true}
  >
    <div
      className="ui centered grid"
    >
      <GridRow>
        <div
          className="row"
        >
          <LoanMetadata
            data={
              Object {
                "availableActions": Object {
                  "action": "action",
                },
                "metadata": Object {
                  "title": "title",
                },
              }
            }
          >
            <div>
              <Header
                as="h1"
              >
                <h1
                  className="ui header"
                >
                  Loan Information
                </h1>
              </Header>
              <Table
                as="table"
                celled={true}
                compact={true}
                definition={true}
                textAlign="center"
              >
                <table
                  className="ui celled definition compact center aligned table"
                >
                  <TableHeader
                    as="thead"
                    fullWidth={true}
                  >
                    <thead
                      className="full-width"
                    >
                      <TableRow
                        as="tr"
                        cellAs="td"
                      >
                        <tr
                          className=""
                        >
                          <TableHeaderCell
                            as="th"
                          >
                            <TableCell
                              as="th"
                              className=""
                            >
                              <th
                                className=""
                              >
                                Property
                              </th>
                            </TableCell>
                          </TableHeaderCell>
                          <TableHeaderCell
                            as="th"
                          >
                            <TableCell
                              as="th"
                              className=""
                            >
                              <th
                                className=""
                              >
                                Value
                              </th>
                            </TableCell>
                          </TableHeaderCell>
                        </tr>
                      </TableRow>
                    </thead>
                  </TableHeader>
                  <TableBody
                    as="tbody"
                  >
                    <tbody
                      className=""
                    >
                      <TableRow
                        as="tr"
                        cellAs="td"
                        key="title"
                        name="loanMetadataRow"
                      >
                        <tr
                          className=""
                          name="loanMetadataRow"
                        >
                          <TableCell
                            as="td"
                            collapsing={true}
                          >
                            <td
                              className="collapsing"
                            >
                              title
                            </td>
                          </TableCell>
                          <TableCell
                            as="td"
                          >
                            <td
                              className=""
                            >
                              title
                            </td>
                          </TableCell>
                        </tr>
                      </TableRow>
                    </tbody>
                  </TableBody>
                </table>
              </Table>
            </div>
          </LoanMetadata>
        </div>
      </GridRow>
      <GridRow>
        <div
          className="row"
        >
          <LoanActions
            actionLoading={false}
            actions={
              Object {
                "action": "action",
              }
            }
            data={
              Object {
                "title": "title",
              }
            }
            onAction={[Function]}
          >
            <List
              horizontal={true}
            >
              <div
                className="ui horizontal list"
                role="list"
              >
                <ListItem
                  key="action"
                >
                  <div
                    className="item"
                    onClick={[Function]}
                    role="listitem"
                  >
                    <Button
                      as="button"
                      onClick={[Function]}
                      primary={true}
                      role="button"
                    >
                      <button
                        className="ui primary button"
                        onClick={[Function]}
                        role="button"
                      >
                        action
                      </button>
                    </Button>
                  </div>
                </ListItem>
              </div>
            </List>
          </LoanActions>
        </div>
      </GridRow>
    </div>
  </Grid>
</LoanDetails>
`;

exports[`tests the LoanDetails component renders the loader component when fetching is in progress 1`] = `
<LoanDetails
  actionLoading={false}
  data={Object {}}
  error={Object {}}
  fetchLoading={true}
  fetchLoanDetails={[Function]}
  loanActionError={false}
  postLoanAction={[Function]}
>
  <Loader
    active={true}
    inline="centered"
  >
    <div
      className="ui active centered inline loader"
    />
  </Loader>
</LoanDetails>
`;
